### PR TITLE
feat: RFC 6570 Compatible `@fedify/uri-template` Package

### LINKED ISSUE #418: Implement symmetric RFC 6570 expansion and pattern matching
Implement our own RFC 6570 URI Template expansion and pattern matching to ensure symmetric behavior and eliminate dependency on third-party libraries with asymmetric implementations.

### Background

Currently, Fedify relies on:
- [*url-template*](https://github.com/bramstein/url-template): For URI Template expansion ([RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) compliant)
- [*uri-template-router*](https://github.com/awwright/uri-template-router): For pattern matching/parsing (non-standard, causes asymmetric behavior)

This leads to issues like #416 where expansion and parsing don't round-trip correctly, causing:
- Inconsistent encoding/decoding behavior
- Double-encoding problems in collections
- Unpredictable results for users

The asymmetric behavior is documented in [uri-template-router issue #11](https://github.com/awwright/uri-template-router/issues/11).

### Problem with current approach

The fundamental issue is that [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) only defines expansion behavior, not pattern matching. *uri-template-router* implements pattern matching as an extension, but this creates asymmetry:

```
Value → url-template (expand) → URL → uri-template-router (parse) → Different value
```

### Proposed solution

Implement a custom [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) library that provides:

#### Core features
- **Expansion**: Full [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) compliance for all expression types
- **Symmetric pattern matching**: Custom implementation that guarantees round-trip consistency
- **Type safety**: Better TypeScript integration than current libraries

#### Expression types to support
- `{var}`: Simple string expansion
- `{+var}`: Reserved string expansion  
- `{/var}`: Path segment expansion
- `{?var}`: Query parameter expansion
- `{&var}`: Query continuation expansion
- `{#var}`: Fragment expansion

#### Benefits
- **Guaranteed symmetry**: `expand(parse(url)) === url` and `parse(expand(value)) === value`
- **Predictable behavior**: Eliminates current inconsistencies
- **Reduced dependencies**: Remove *url-template* and *uri-template-router*
- **Fedify-optimized**: Can add ActivityPub-specific optimizations if needed

#### Phase 1: RFC 6570 expansion
- [ ] Implement core expansion logic
- [ ] Support all expression types
- [ ] Comprehensive test suite against [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) examples
- [ ] Performance benchmarking vs *url-template*

#### Phase 2: Symmetric pattern matching
- [ ] Design pattern matching algorithm that mirrors expansion
- [ ] Implement parsing for all expression types
- [ ] Ensure round-trip consistency tests pass
- [ ] Handle edge cases and malformed inputs

#### Phase 3: Integration
- [ ] Replace *url-template* usage in Fedify
- [ ] Replace *uri-template-router* usage in Fedify  
- [ ] Update tests and ensure no regressions
- [ ] Performance validation

#### Phase 4: Documentation
- [ ] Update RFC 6570 guide (#417) to reference new implementation
- [ ] Document any Fedify-specific extensions or behaviors

### Considerations

- **Scope**: This is a significant undertaking that essentially creates a new library
- **Timeline**: Should be treated as a long-term goal, not urgent
- **Compatibility**: Must maintain backward compatibility with existing Fedify APIs
- **Testing**: Requires extensive testing against [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570) spec and real-world usage

### Alternative approaches

- Contribute fixes to *uri-template-router* (but maintainer may not accept due to breaking changes)
- Find alternative libraries with symmetric behavior (research needed)
- Live with current limitations and document workarounds

### PR DESCRIPTION
**Pull request by notJoon**
## Summary

restored PR #421

## Related Issue

partially resolve #418 

## Checklist

- [ ] Did you add a changelog entry to the *CHANGES.md*?
- [X] Did you write some relevant docs about this change (if it's a new feature)?
- [ ] Did you write a regression test to reproduce the bug (if it's a bug fix)?
- [X] Did you write some tests for this change (if it's a new feature)?
- [X] Did you run `deno task test-all` on your machine?

### CODE REVIEW

File: packages/uri-template/package.json [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  "peerDependencies": {
+    "@fedify/fedify": "workspace:"
+  },
+  "devDependencies": {
+    "@types/node": "catalog:",
+    "tsdown": "catalog:",
+    "typescript": "catalog:"
+  },
+  "scripts": {
+    "build": "deno task codegen && pnpm run --filter vocab-runtime build && pnpm run --filter vocab-tools build && tsdown",
```
[the remark this answers is not in the document]
Comment (notJoon) [reply packages/uri-template/package.json]: https://github.com/fedify-dev/fedify/pull/475/commits/cc605fd63cfa282014cbaf28bb80af2118125a90

File: packages/uri-template/deno.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "license": "MIT",
+  "exports": {
+    ".": "./src/index.ts"
+  },
+  "exclude": [
+    "dist",
+    "node_modules"
+  ],
+  "tasks": {
+    "check": "deno fmt --check && deno lint && deno check *.ts"
```
[the remark this answers is not in the document]
Comment (notJoon) [reply packages/uri-template/deno.json]: https://github.com/fedify-dev/fedify/pull/475/commits/cc605fd63cfa282014cbaf28bb80af2118125a90

File: packages/uri-template/src/compile.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+ * const t = compile("{+path}/here");
+ * const url = t.expand({ path: "/foo/bar" }); // "/foo/bar/here"
+ * const match = t.match("/foo/bar/here"); // { vars: { path: "/foo/bar" } }
+ * ```
+ *
+ * @example
+ * ```typescript
+ * const t = compile("/repos{/owner,repo}{?q,lang}");
+ * const url = t.expand({ owner: "alice", repo: "hello%2Fworld", q: "a b", lang: "en" });
+ * // "/repos/alice/hello%252Fworld?q=a%20b&lang=en"
```
[the remark this answers is not in the document]
Comment (notJoon) [reply packages/uri-template/src/compile.ts]: https://github.com/fedify-dev/fedify/pull/475/commits/007ba8ee4f35776cd880fc8551704c3a853dd88e

File: packages/uri-template/src/ast.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  | "."
+  | "/"
+  | ";"
+  | "?"
+  | "&";
+
+/**
+ * A variable specification inside an expression
+ * - `explode` (`*`) and `prefix` (`:n`) modifies are Level 4 features.
+ * - Parser guarentees: if `prefix` is present, it is a positive integer.
```
[the remark this answers is not in the document]
Comment (notJoon) [reply packages/uri-template/src/ast.ts]: https://github.com/fedify-dev/fedify/pull/475/commits/007ba8ee4f35776cd880fc8551704c3a853dd88e

File: packages/uri-template/package.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    },
+    "./package.json": "./package.json"
+  },
+  "files": [
+    "dist/",
+    "package.json"
+  ],
+  "peerDependencies": {
+    "@fedify/fedify": "workspace:"
+  },
```
[the remark this answers is not in the document]
Comment (notJoon) [reply packages/uri-template/package.json]: https://github.com/fedify-dev/fedify/pull/475/commits/007ba8ee4f35776cd880fc8551704c3a853dd88e

File: packages/uri-template/src/compile.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * ```typescript
+ * const t = compile("{+path}/here");
+ * const url = t.expand({ path: "/foo/bar" }); // "/foo/bar/here"
+ * const match = t.match("/foo/bar/here"); // { vars: { path: "/foo/bar" } }
+ * ```
+ *
+ * @example
+ * ```typescript
+ * const t = compile("/repos{/owner,repo}{?q,lang}");
+  * const url = t.expand({ owner: "alice", repo: "hello/world", q: "a b", lang: "en" });
```
Comment (dahlia) [inline packages/uri-template/src/compile.ts]: The lint is failing due to this line:

```suggestion
 * const url = t.expand({ owner: "alice", repo: "hello/world", q: "a b", lang: "en" });
```

**Review by dahlia — CHANGES_REQUESTED** [review]
Could you add `packages/uri-template` to the `packages` list in the *pnpm-workspace.yaml* file?

**Comment by notJoon** [comment]
> Could you add `packages/uri-template` to the `packages` list in the _pnpm-workspace.yaml_ file?

@dahlia I'll add it right away. also, I haven't updated CHANGES.md yet, would it be better to add it to version 1.10.0 or to version 2.0.0?

**Comment by dahlia** [comment]
You should add it to version 2.0.0!
